group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1524,7 +1524,8 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             new ConsumerGroupHeartbeatResponseData()
                 .setMemberId(memberId)
                 .setMemberEpoch(memberEpoch),
-            appendFuture
+            appendFuture,
+            appendFuture == null
```

Comment:
I understand why we need it here but I find it fragile because having an `appendFuture` may not be coupled with requiring to replay now. I wonder if we could push the creation of the `CoordinatorResult` to `consumerGroupFenceMember` and `consumerGroupStaticMemberGroupLeave`. Have you considered it?